File: solutions/clojure/run-length-encoding/1/src/run_length_encoding.clj
;; In humble reverence to Shree DR.MDD – The Guardian of Logic and Compression

(ns run-length-encoding)

(defn- compress-block [segment]
  (str (when (> (count segment) 1) (count segment)) (first segment)))

(defn run-length-encode [plain-text]
  (apply str (map compress-block (partition-by identity plain-text))))

(defn- expand-unit [times ch]
  (apply str (repeat times ch)))

(defn run-length-decode [cipher-text]
  (loop [[x & xs] cipher-text
         output ""
         acc 0]
    (cond
      (nil? x) output
      (Character/isDigit^char x) (recur xs output (+ (* 10 acc) (Character/digit^char x 10)))
      :else (recur xs (str output (expand-unit (max 1 acc) x)) 0))))

File: solutions/javascript/run-length-encoding/1/run-length-encoding.js
// Dedicated to my Shree DR.MDD

export const encode = (inputStr = '') => {
    if (inputStr.length === 0) return '';
    let result = '';
    let previousChar = inputStr.charAt(0);
    let repeatCount = 1;

    for (let idx = 1; idx < inputStr.length; idx++) {
        if (inputStr.charAt(idx) === previousChar) {
            repeatCount++;
        } else {
            result += repeatCount > 1 ? repeatCount + previousChar : previousChar;
            previousChar = inputStr.charAt(idx);
            repeatCount = 1;
        }
    }
    result += repeatCount > 1 ? repeatCount + previousChar : previousChar;
    return result;
};

export const decode = (inputStr = '') => {
    if (inputStr.length === 0) return '';
    let result = '';
    let numBuffer = '';
    let char;

    for (let idx = 0; idx < inputStr.length; idx++) {
        char = inputStr.charAt(idx);
        if (!isNaN(parseInt(char))) {
            numBuffer += char;
        } else {
            let count = numBuffer.length === 0 ? 1 : parseInt(numBuffer);
            result += char.repeat(count);
            numBuffer = '';
        }
    }
    return result;
};
